### PR TITLE
[HLSL] Implement `WaveActiveAllTrue` Intrinsic

### LINKED ISSUE #99161: Implement the `WaveActiveAllTrue` HLSL Function
- [ ] Implement `WaveActiveAllTrue` clang builtin,
- [ ] Link `WaveActiveAllTrue` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `WaveActiveAllTrue` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `WaveActiveAllTrue` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/WaveActiveAllTrue.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/WaveActiveAllTrue-errors.hlsl`
- [ ] Create the `int_dx_WaveActiveAllTrue` intrinsic in `IntrinsicsDirectX.td`
- [ ] Create the `DXILOpMapping` of `int_dx_WaveActiveAllTrue` to  `114` in `DXIL.td`
- [ ] Create the  `WaveActiveAllTrue.ll` and `WaveActiveAllTrue_errors.ll` tests in `llvm/test/CodeGen/DirectX/`
- [ ] Create the `int_spv_WaveActiveAllTrue` intrinsic in `IntrinsicsSPIRV.td`
- [ ] In SPIRVInstructionSelector.cpp create the `WaveActiveAllTrue` lowering and map  it to `int_spv_WaveActiveAllTrue` in `SPIRVInstructionSelector::selectIntrinsic`.
- [ ] Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll`

## DirectX

| DXIL Opcode | DXIL OpName | Shader Model | Shader Stages |
| ----------- | ----------- | ------------ | ------------- |
| 114 | WaveAllTrue | 6.0 | ('library', 'compute', 'amplification', 'mesh', 'pixel', 'vertex', 'hull', 'domain', 'geometry', 'raygeneration', 'intersection', 'anyhit', 'closesthit', 'miss', 'callable', 'node') |

## SPIR-V

# [OpGroupNonUniformAll](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll):

## Description:
  
Evaluates a predicate for all active invocations in the [group](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Group),
resulting in **true** if predicate evaluates to **true** for all active
[invocations](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Invocation) in the group, otherwise the result is
**false**.  
  
*Result Type* must be a [*Boolean type*](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Boolean).  
  
*Execution* is a [*Scope*](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Scope_-id-) that identifies the group of
invocations affected by this command. It must be **Subgroup**.  
  
*Predicate* must be a [*Boolean type*](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Boolean).

[Capability](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Capability):  
**GroupNonUniformVote**  
  
[Missing before](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Unified) **version 1.3**.

<table style="width:100%;">
<colgroup>
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 16%" />
</colgroup>
<thead>
<tr>
<th>Word Count</th>
<th>Opcode</th>
<th>Results</th>
<th>Operands</th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p>5</p></td>
<td class="tableblock halign-left valign-top"><p>334</p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>Result Type</em></p></td>
<td class="tableblock halign-left valign-top"><p><a
href="#ResultId"><em>Result &lt;id&gt;</em></a></p></td>
<td class="tableblock halign-left valign-top"><p><a
href="#Scope_-id-"><em>Scope &lt;id&gt;</em></a><br />
<em>Execution</em></p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>Predicate</em></p></td>
</tr>
</tbody>
</table>

### Example 1
```hlsl
//dxc WaveActiveAllTrue_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export bool fn(bool p1) {
    return WaveActiveAllTrue(p1);
}
```
## HLSL:

Returns true if the expression is true in all active lanes in the current wave.

## Syntax

``` syntax
bool WaveActiveAllTrue(
   bool expr
);
```

## Parameters

<dl> <dt>

*expr* 
</dt> <dd>

The boolean expression to evaluate.

</dd> </dl>

## Return value

True if the expression is true in all lanes.

## Remarks

This function is supported from shader model 6.0 in all shader stages. 

 

## See also

<dl> <dt>

[Overview of Shader Model 6](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/hlsl-shader-model-6-0-features-for-direct3d-12.md)
</dt> <dt>

[Shader Model 6](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/shader-model-6-0.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by V-FEXrt**
Resolves https://github.com/llvm/llvm-project/issues/99161

- [x]  Implement `WaveActiveAllTrue` clang builtin,
- [x]  Link `WaveActiveAllTrue` clang builtin with `hlsl_intrinsics.h`
- [x]  Add sema checks for `WaveActiveAllTrue` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [x]  Add codegen for `WaveActiveAllTrue` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [x]  Add codegen tests to `clang/test/CodeGenHLSL/builtins/WaveActiveAllTrue.hlsl`
- [x]  Add sema tests to `clang/test/SemaHLSL/BuiltIns/WaveActiveAllTrue-errors.hlsl`
- [x]  Create the `int_dx_WaveActiveAllTrue` intrinsic in `IntrinsicsDirectX.td`
- [x]  Create the `DXILOpMapping` of `int_dx_WaveActiveAllTrue` to `114` in `DXIL.td`
- [x]  Create the `WaveActiveAllTrue.ll` and `WaveActiveAllTrue_errors.ll` tests in `llvm/test/CodeGen/DirectX/`
- [x]  Create the `int_spv_WaveActiveAllTrue` intrinsic in `IntrinsicsSPIRV.td`
- [x]  In SPIRVInstructionSelector.cpp create the `WaveActiveAllTrue` lowering and map it to `int_spv_WaveActiveAllTrue` in `SPIRVInstructionSelector::selectIntrinsic`.
- [x]  Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll`

### CODE REVIEW

File: llvm/lib/Target/DirectX/DXIL.td [hunk 1]
```diff
@@ -861,6 +861,14 @@ def CreateHandleFromBinding : DXILOp<217, createHandleFromBinding> {
   let stages = [Stages<DXIL1_6, [all_stages]>];
 }
 
+def WaveActiveAllTrue : DXILOp<334, waveAllTrue> {
```
Comment (V-FEXrt) [inline llvm/lib/Target/DirectX/DXIL.td:864]: Crossed wires with SPIRV. This should be 114  not 334

File: llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+; CHECK-DAG: %[[#uint:]] = OpTypeInt 32 0
+; CHECK-DAG: %[[#scope:]] = OpConstant %[[#uint]] 3
+; CHECK-DAG: OpCapability GroupNonUniformVote
+
+; CHECK-LABEL: Begin function test_wave_all
+define i1 @test_wave_all(i1 %p1) #0 {
+entry:
+; CHECK: %[[#param:]] = OpFunctionParameter %[[#bool]]
+; CHECK: %{{.+}} = OpGroupNonUniformAll %[[#bool]] %[[#scope]] %[[#param]]
+  %0 = call token @llvm.experimental.convergence.entry()
```
Comment (inbelic) [inline llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll:14]: I think the convergent semantics are redundant here

**Review by inbelic — APPROVED** [review]
LGTM, just a nit and your self-review

File: llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll (hunk 2 above)
Comment (V-FEXrt) [reply llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll:14]: @Keenuts requested they be included here https://github.com/llvm/llvm-project/pull/115902#discussion_r1850647413

Comment (inbelic) [reply llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveActiveAllTrue.ll:14]: Ah okay, I thought that comment had just been for checking that they are emitted. But I guess it doesn't hurt to add them.

**Review by spall — APPROVED** [review]
LGTM

**Pull request merged**